src/xarray_regrid/methods/conservative.py

```diff
@@ -102,132 +114,88 @@ def conservative_regrid_dataset(
+                data_vars[array], valid_fracs[array] = apply_weights(
+                    data_vars[array], weights, coord, valid_fracs[array], skipna
+                )
+                # Mask out any regridded points outside the original domain
+                data_vars[array] = data_vars[array].where(covered_grid)
 
-    for coord in coords:
-        uncovered_target_grid = (coords[coord] <= data[coord].max()) & (
-            coords[coord] >= data[coord].min()
-        )
+    if skipna:
+        # Mask out any points that don't meet the nan threshold
+        valid_threshold = get_valid_threshold(nan_threshold)
+        for array, da in data_vars.items():
+            data_vars[array] = da.where(valid_fracs[array] >= valid_threshold)
 
-        if coord in data.coords:
-            target_coords = coords[coord].to_numpy()
-            source_coords = data[coord].to_numpy()
+    for array, attrs in data_attrs.items():
+        data_vars[array].attrs = attrs
 
-            weights = get_weights(source_coords, target_coords)
+    ds_regridded = xr.Dataset(data_vars=data_vars, attrs=ds_attrs)
 
-            # Modify weights to correct for latitude distortion
-            if str(coord) == latitude_coord:
-                dot_array = utils.create_dot_dataarray(
-                    weights, str(coord), target_coords, source_coords
-                )
-                dot_array = apply_spherical_correction(dot_array, latitude_coord)
-                weights = dot_array.to_numpy()
+    for coord, attrs in coord_attrs.items():
+        if coord not in ds_regridded.coords:
+            # Add back any additional coordinates from the original dataset
+            ds_regridded[coord] = data_coords[coord]
+        ds_regridded[coord].attrs = attrs
 
-            data = data.transpose(coord, ...)
-            data = apply_weights(data, weights, coord, target_coords)
+    return ds_regridded
 
-            # Replace zeros outside of original data grid with NaNs
-            data = data.where(uncovered_target_grid)
 
-    new_coords = [data[coord] for coord in data_coords]
-    for coord, attr in zip(new_coords, coord_attrs, strict=True):
-        coord.attrs = attr
-    data.attrs = attrs
+def apply_weights(
+    da: xr.DataArray,
+    weights: np.ndarray,
+    coord_name: Hashable,
+    valid_frac: xr.DataArray | None,
+    skipna: bool,
+) -> tuple[xr.DataArray, xr.DataArray]:
+    """Apply the weights to convert data to the new coordinates."""
+    coord_map = {f"target_{coord_name}": coord_name}
+    weights_norm = weights.copy()
 
-    return data
+    if skipna:
+        notnull = da.notnull()
+        da = da.fillna(0)
+        # Renormalize the weights along this dim by the accumulated valid_frac
+        # along previous dimensions
+        if valid_frac is not None:
+            weights_norm = weights * valid_frac / valid_frac.mean(coord_name)
 
+    da_reduced = xr.dot(da, weights_norm, dim=coord_name, optimize=True)
```

Comment:
I had used np/xr.einsum originally because I did find a performance difference. But perhaps that's gone due to changes in the code since then, or elsewhere. Just using `xr.dot` simplifies things!